lib/stores/filesystem.js

```diff
@@ -0,0 +1,382 @@
+          metadata["x-amz-meta-" + key] = value;
+        }
+
+        done(null, metadata);
+      }
+    );
+  }
+
+  storeMetadata(bucket, key, metadata, md5, done) {
+    if (typeof md5 === "function") {
+      done = md5;
+      md5 = null;
+    }
+
+    const metadataFile = path.join(
+      this.getBucketPath(bucket),
+      key,
+      METADATA_FILE
+    );
+    const md5File = path.join(
+      this.getBucketPath(bucket),
+      key,
+      CONTENT_FILE + ".md5"
+    );
+
+    // this is only for backwards compatibility purposes
+    // otherwise we can just dump the result of JSON.stringify(metadata) into the file
+
+    const json = {
+      cacheControl: metadata["cache-control"],
+      contentDisposition: metadata["content-disposition"],
+      contentEncoding: metadata["content-encoding"],
+      contentLanguage: metadata["content-language"],
+      contentType: metadata["content-type"],
+      expires: metadata["expires"],
+      websiteRedirectLocation: metadata["website-redirect-location"]
+    };
+    json.customMetaData = entries(metadata)
+      .filter(([header]) => header.startsWith("x-amz-meta-"))
+      .map(([header, value]) => [header.replace("x-amz-meta-", ""), value]);
+
+    if (md5) {
+      fs.writeFileSync(md5File, md5);
+    }
+    fs.writeFile(metadataFile, JSON.stringify(json), done);
+  }
+
+  // store implementation
+
+  getBuckets() {
+    const buckets = [];
+    for (const filename of fs.readdirSync(this.rootDirectory)) {
+      const file = fs.statSync(path.resolve(this.rootDirectory, filename));
+      if (file.isDirectory()) {
+        buckets.push(new Bucket(filename, file.ctime));
+      }
+    }
+    return buckets;
+  }
+
+  getBucket(bucket, done) {
+    const bucketPath = this.getBucketPath(bucket);
+    fs.stat(bucketPath, (err, file) => {
+      if (err || !file.isDirectory()) {
+        return done("Bucket not found");
+      }
+      done(null, new Bucket(bucket, file.ctime));
+    });
+  }
+
+  putBucket(bucket, done) {
+    const bucketPath = this.getBucketPath(bucket);
+    fs.mkdirp(bucketPath, 502, err => {
+      if (err) return done(err);
+      this.getBucket(bucket, done);
+    });
+  }
+
+  deleteBucket(bucket, done) {
+    fs.rmdir(this.getBucketPath(bucket), done);
+  }
+
+  listObjects(bucket, options, done) {
+    const bucketPath = this.getBucketPath(bucket);
+    const commonPrefixes = new Set();
+    let keys = utils
+      .walk(bucketPath)
+      .map(key => key.slice(bucketPath.length + 1));
+
+    if (!keys.length) {
+      return done(null, {
+        objects: [],
+        commonPrefixes: [],
+        isTruncated: false
+      });
+    }
+
+    if (options.prefix) {
+      keys = keys.filter(key => key.startsWith(options.prefix));
+    }
+
+    if (options.delimiter) {
+      const prefix = options.prefix || "";
+      keys = keys.filter(key => {
+        const idx = key.slice(prefix.length).indexOf(options.delimiter);
+        if (!~idx) return true;
```

Comment:
I had to Google this `!~idx` to understand what you were trying to do. Pretty clever, but think there are more readable ways to express this. [`String.prototype.includes`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes) or [`lodash.includes`](https://lodash.com/docs/4.17.5#includes) would make this logic look closer to English.

Reply:
I guess it might be a bit obscure but I've found it to be pretty readable once you're used to it. (If you want something really funky look up using `~~( )` as an efficient floor function!)

I'll just convert it to `idx !== -1` to help readability. I don't really want to do `includes` since I need to know the index if it does exist.